haskell-src/Concordium/Types/Parameters.hs

```diff
@@ -687,34 +1404,184 @@ putPoolParameters PoolParametersV1{..} = do
+    get = do
+        tpTimeoutBase <- get
+        tpTimeoutIncrease <- get
+        unless (tpTimeoutIncrease > 1) $ fail "timeoutIncrease must be greater than 1."
+        tpTimeoutDecrease <- get
+        unless (tpTimeoutDecrease > 0) $ fail "timeoutDecrease must be greater than 0."
+        unless (tpTimeoutDecrease < 1) $ fail "timeoutDecrease must be less than 1."
+        return TimeoutParameters{..}
+
+instance ToJSON TimeoutParameters where
+    toJSON TimeoutParameters{..} =
+        object
+            [ "timeoutBase" AE..= tpTimeoutBase,
+              "timeoutIncrease" AE..= tpTimeoutIncrease,
+              "timeoutDecrease" AE..= tpTimeoutDecrease
+            ]
+
+instance FromJSON TimeoutParameters where
+    parseJSON = withObject "TimeoutParameters" $ \o -> do
+        tpTimeoutBase <- o .: "timeoutBase"
+        tpTimeoutIncrease <- o .: "timeoutIncrease"
+        unless (tpTimeoutIncrease > 1) $ fail "timeoutIncrease must be greater than 1."
+        tpTimeoutDecrease <- o .: "timeoutDecrease"
+        unless (tpTimeoutDecrease > 0) $ fail "timeoutDecrease must be greater than 0."
+        unless (tpTimeoutDecrease < 1) $ fail "timeoutDecrease must be less than 1."
+        return TimeoutParameters{..}
+
+instance HashableTo Hash.Hash TimeoutParameters where
+    getHash = Hash.hash . encode
+
+instance (Monad m) => MHashableTo m Hash.Hash TimeoutParameters
+
+-- * Consensus parameters
+
+-- |Constraint on a type level 'ConsensusParametersVersion' that can be used to get a corresponding
+-- 'SConsensusParametersVersion'. An alias for 'SingI'.
+type IsConsensusParametersVersion (cpv :: ConsensusParametersVersion) = SingI cpv
+
+-- |Witness an 'IsConsensusParametersVersion' constraint for an 'SChainParametersVersion'.
+withIsConsensusParametersVersionFor :: SChainParametersVersion cpv -> (IsConsensusParametersVersion (ConsensusParametersVersionFor cpv) => a) -> a
+withIsConsensusParametersVersionFor scpv = withSingI (sConsensusParametersVersionFor scpv)
+
+-- |Consensus-specific parameters.
+data ConsensusParameters' (cpv :: ConsensusParametersVersion) where
+    ConsensusParametersV0 ::
+        { -- |Election difficulty parameter.
+          _cpElectionDifficulty :: !ElectionDifficulty
+        } ->
+        ConsensusParameters' 'ConsensusParametersVersion0
+    ConsensusParametersV1 ::
+        { -- |Parameters controlling round timeouts.
+          _cpTimeoutParameters :: !TimeoutParameters,
+          -- |Minimum time interval between blocks.
+          _cpMinBlockTime :: !Duration,
+          -- |Maximum energy allowed per block.
+          _cpBlockEnergyLimit :: !Energy
+        } ->
+        ConsensusParameters' 'ConsensusParametersVersion1
+
+-- |Convenience type for a 'ConsensusParameters'' parametrised by the 'ChainParametersVersion'.
+type ConsensusParameters (cpv :: ChainParametersVersion) =
+    ConsensusParameters' (ConsensusParametersVersionFor cpv)
+
+-- |Lens for '_cpElectionDifficulty'
```

Comment:
Could we try to explain things in non-jargon words. It is clear taht this is a Lens. That's what the type says.

But if this comment was to be useful it needs to add something so that non-experts can also understand.

Reply:
Added a comment.